bacula/manpages/bpluginfo.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH bpluginfo "8" "July 2012" "bpluginfo" "Network backup, recovery and verification"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH NAME
bpluginfo \- Bacula Plugin information utility
.SH SYNOPSIS
.B bplufinfo
.RI [ options ]
.I plugin_filename.so
.br
.SH DESCRIPTION
.LP
The main purpose of
.B bpluginfo
is to display different information about Bacula plugin. You can use it to
check a plugin name, author, license and short description. You can use
'-f' option to display API implemented by the plugin. Some plugins may require
additional '-a' option for validating a Bacula Daemons API. In most cases it
is not required.
.PP
./ Bacula is a set of programs for performing a
./ .PP
./ -
./ .BR bpluginfo

.PP
.SH OPTIONS
A summary of options is included below.
.TP
.B \-h
Show usage of the program.
.TP
.BI \-v
Verbose information printing all available data from the plugin, including
plugin header and implemented API.
.TP
.BI \-i
Display short information from plugin header only. This is a default option.
Option incompatible with
.B -f
option.
.TP
.BI \-f
Display information about implemented API functions.
.TP
.BI \-a\ <api>
You can supply the plugin initialization function with a particular Bacula
API number. Without this option a default API number is '1'. Option require
a numeric argument.
.SH RETURN CODE
.BR bpluginfo
returns 0 on success, and non-zero on error.
.TP
You can check return code to find what was a cause of the error.
 * 0 - success
 * 1 - cannot load a plugin
 * 2 - cannot find a loadPlugin function
 * 3 - cannot find an unloadPlugin function
 * 10 - not enough memory
.SH EXAMPLE USAGE
This is an example of bplufinfo usage with verbose option (-v) and default plugin.
.LP
.sp
.RS
.nf

\fB$ bpluginfo -v bpipe-fd.so

Plugin type:            File Daemon plugin
Plugin magic:           *FDPluginData*
Plugin version:         1
Plugin release date:    January 2008
Plugin author:          Kern Sibbald
Plugin license:         Bacula AGPLv3
Plugin description:     Bacula Pipe File Daemon Plugin
Plugin API version:     6

Plugin functions:
 newPlugin()
 freePlugin()
 getPluginValue()
 setPluginValue()
 handlePluginEvent()
 startBackupFile()
 endBackupFile()
 startRestoreFile()
 endRestoreFile()
 pluginIO()
 createFile()
 setFileAttributes()
.fi
.RE

.SH AUTHOR
Written by Radoslaw Korzeniewski (c) Inteos Sp. z o.o.
.SH BUGS
Does not handle all required bacula functions callbacks which can lead to
utility crash.
.\".SH TODO"

.PP

.PP

.SH "REPORTING BUGS"
Report bugs to <support@inteos.pl>.
.SH COPYRIGHT
Copyright \(co 2012 Free Software Foundation Europe e.V.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR bacula-dir,
.BR bacula-sd,
.BR bacula-fd,
.BR "Bacula Plugins API"


bacula/ChangeLog
              Changelog on version 7.0.2

Release version 7.0.2
02Apr14
 - Remove more vestiges of libbacpy
 - Put back @PYTHON@ path in configure
 - Fix improper string in parser
 - Remove libbacpy from rpm spec files
 - Fix linking check_bacula
 - Fix new SD login in check_bacula
 - Tweek docs build process

Release version 7.0.1
31Mar14
 - Remove old plugin-test
 - Update po files
 - Enable installation of the bpluginfo utility
 - More tray-monitor updates
 - Add Simone Caronii to AUTHORS
 - Align command line switches in manpages.
 - Apply upgrade to config.guess
 - Remove bgnome-console and bwx-console leftovers.
 - Update tray-monitor header also for new bsock calls
 - Attempt to fix nagios to use new bsock calls
 - Update tray-monitor to new bsock calls

==========
Release version 7.0.0

24Mar14
 - Add Josip Almasi to AUTHORS
 - [PATCH] Support for restricted consoles in BAT config
 - [PATCH] Fix for free director directive
 - [PATCH] Fix auto-select restricted console for director in bconsole
 - Realign output display
 - Update ua_output.c from Branch-6.7
 - Add some missing Branch-6.7 updates
 - Added needed empty directories to Baculum
 - Fix for support PostgreSQL, MySQL and SQLite
 - Framework adjusting to Baculum database connections
 - Framework fix for lower case tables names in MySQL
 - Fix for Baculum SQLite support
 - Initial commit Baculum
 - Add Marcin to AUTHORS file
 - Strip trailing blanks
 - Update copyright year
 - Update LICENSE and header files
 - Remove old file
 - Add new header in misc files
 - Remove tray-monitor bwx-console manual installation
 - Remove FD python and examples
 - Fixup spec files
 - Remove pythonlib from lib
 - Update package-list
 - Fix SDCallsClient daemon synchronization
 - Add debug code + make 127.0.0.1 same as localhost for tls tests
 - Fix multiple DIRs in console
 - Make failure for bat to connect to DIR non-fatal
 - Fix bat style to one that works
 - Take disk-changer from Branch-6.7
 - Simplify Version output
 - Fix FDVersion for SD Calls Client test
 - Update accurate test
 - Update differential test
 - Add new regress timing scripts
 - Improve plugin make clean
 - Implement regress FORCE_SDCALLS
 - Remove win32 tray-monitor and wx-console directories
 - Remove regress-config need only regress-config.in
 - Add configure archivedir
 - Improve SQL failure reporting
 - Major backport from BEE to community
 - Add copyright to mtx-changer.in

Release version 5.2.13

19Feb13
 - Fix build/configure problems with bpluginfo.c
 - Add missing vol_mgr.h file
 - Add timer to run_multiple

18Feb13
 - Refactor lock_volumes so most lock a vol rather than globally
 - Add virtualfull-extreme test

17Feb13
 - Apply patch for chio-changer-openbsd from bug #1984 -- Implements listall
 - Add bat Mac patch from bug #1953

16Feb13
 - Fix bug #1812 cannot run Copy/Migrate jobs from bat
 - Fix loading of bat translations. Patch from bug #1890
 - Fix text input in bat. Fixe bug #1965
 - Remove Device from show command as it is not used or updated

13Feb13
 - Fix #1982 update enabled keyword in help command

08Feb13
 - Fix update_postgresql_tables 10 to 11

30Jan13
 - Clean after building doc binaries requested by Willem vd Akker for Debian
  build

29Jan13
 - Remove unused method
 - Require Qt version 4.8.4
 - Fix bug #1955 that OK to run? does not retry on bad response
 - Backport new lock calls + debug for SD

26Jan13
 - Remove old todo

09Jan13
 - Fix bug #1975 new label ignored when first is duplicate.

31Dec12
 - Require Qt 4.8.1 to build bat

27Dec12
 - Make compile of timelimit a bit more portable

26Dec12
 - Tweek turn off prune-test
 - Pull src/lib changes from master
 - Pull SD files from master

25Dec12
 - Turn off prune-test
 - Add logdir and bsrdir to regression config
 - Add bsrdir and logdir to regress-config
 - Rebuild configure
 - Changed default directories from /var/bacula to /opt/bacula
 - Fix patches from master
 - Make spool size max message more explicit
 - Display more info when maximum spool size is reached

23Dec12
 - Fix TERM defs for Solaris 11 in conio.c

21Dec12
 - Force DEVELOPER on during regression setup

14Dec12
 - Allow group to cd to sysconfdir

09Dec12
 - Make sysconfdir belong to Bacula user/group and exclude others

07Dec12
 - Fix Virtual Full file close bug -- race condition
 - Enhance mount message to include read/append

06Dec12
 - Improve vol_mgr swap messages
 - Add Jmsg7 and Jmsg8
 - Enhance output of Using Device to include for read/write

04Dec12
 - Add jobid to Pmsg output

29Nov12
 - Fix bug #1959 input validation on delete of jobs.

24Nov12
 - Fix bug #1956 Authorization Errors false

23Nov12
 - Add workaround for #5507 where autoprune and reload are in deadlock

21Nov12
 - Allow conf file quoted strings to be used in a list
 - Make bfgets handle very long lines
 - Backport from Enterprise

17Nov12
 - Avoid seg fault by checking for NULL client -- reported by Arno

10Nov12
 - Final fix of bug #1943

08Nov12
 - Fix bug #1948 MailOnSuccess get executed on error.

07Nov12
 - Fix missing index on Media table

06Nov12
 - Fix bug #1943 no message storage on closed database connection.
 - Fix bug #1946 Fix problem with MySQL with big Base jobs.

02Nov12
 - Reduce wait time in regression RunScripts
 - Attempt to indicate timed out tests
 - Implement a timeout on regression tests of 15 minutes
 - Fix #5346 .bvfs_lsfiles and .bvfs_restore to handle deleted files

01Nov12
 - Fix spooldata, accurate and ingnoreduplicate run argument

21Oct12
 - Add remote regress test script

16Oct12
 - Fix messages segfault

10Oct12
 - Recompile configure.in
 - Implement feature request #1939

07Oct12
 - fix #1938 about PATH_MAX on hurd
 - fix #1938 about MAXPATHLEN on hurd

05Oct12
 - Fix bug #1937 OpenBSD autochanger example script

25Sep12
 - Fix #4996 about MaxRunTime canceling the job too early

19Sep12
 - Fix bug #1934 Wrong values at media list in BAT

18Sep12
 - Fix race condition in close_msg that causes seg fault

16Sep12

bacula/src/tools/bpluginfo.c
/*
 * Contributed in 2012 by Inteos sp. z o.o.
 *
 * Utility tool display various information about Bacula plugin,
 * including but not limited to:
 * - Name and Author of the plugin
 * - Plugin License
 * - Description
 * - API version
 * - Enabled functions, etc.
 */
/*
   Bacula® - The Network Backup Solution

   Copyright (C) 2006-2014 Free Software Foundation Europe e.V.

   The main author of Bacula is Kern Sibbald, with contributions from many
   others, a complete list can be found in the file AUTHORS.

   You may use this file and others of this release according to the
   license defined in the LICENSE file, which includes the Affero General
   Public License, v3.0 ("AGPLv3") and some additional permissions and
   terms pursuant to its AGPLv3 Section 7.

   Bacula® is a registered trademark of Kern Sibbald.
*/

#include <stdio.h>
#include <unistd.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#ifndef __WIN32__
#include <dlfcn.h>
#endif
#include "bacula.h"
#include "../filed/fd_plugins.h"
#include "../dird/dir_plugins.h"
// I can't include sd_plugins.h here ...
#include "../stored/stored.h"
#include "assert_macro.h"

extern "C" {
   typedef int (*loadPlugin) (void *binfo, void *bfuncs, void **pinfo,
               void **pfuncs);
   typedef int (*unloadPlugin) (void);
}
#define DEFAULT_API_VERSION   1
enum plugintype {
   DIRPLUGIN,
   FDPLUGIN,
   SDPLUGIN,
   ERRORPLUGIN,
};

/*
 * pDirInfo
 * pInfo
 * psdInfo
 */
typedef union _pluginfo pluginfo;
union _pluginfo {
   pDirInfo pdirinfo;
   pInfo pfdinfo;
   psdInfo psdinfo;
};

/*
 * pDirFuncs
 * pFuncs
 * psdFuncs
 */
typedef union _plugfuncs plugfuncs;
union _plugfuncs {
   pDirFuncs pdirfuncs;
   pFuncs pfdfuncs;
   psdFuncs psdfuncs;
};

/*
 * bDirFuncs
 * bFuncs
 * bsdFuncs
 */
/*
 * TODO: change to union
 *
typedef union _baculafuncs baculafuncs;
union _baculafuncs {
   bDirFuncs bdirfuncs;
   bFuncs bfdfuncs;
   bsdFuncs bsdfuncs;
};
*/
typedef struct _baculafuncs baculafuncs;
struct _baculafuncs {
   uint32_t size;
   uint32_t version;
   int (*registerBaculaEvents) (void *ctx, ...);
   int (*getBaculaValue) (void *ctx, int var, void *value);
   int (*setBaculaValue) (void *ctx, int var, void *value);
   int (*JobMessage) (void *ctx, const char *file, int line, int type, int64_t mtime,
            const char *fmt, ...);
   int (*DebugMessage) (void *ctx, const char *file, int line, int level,
         const char *fmt, ...);
   void *(*baculaMalloc) (void *ctx, const char *file, int line, size_t size);
   void (*baculaFree) (void *ctx, const char *file, int line, void *mem);
};

/*
 * bDirInfo
 * bInfo
 * bsdInfo
 */
typedef union _baculainfos baculainfos;
union _baculainfos {
   bDirInfo bdirinfo;
   bInfo bfdinfo;
   bsdInfo bsdinfo;
};

/*
typedef struct _baculainfos baculainfos;
struct _baculainfos {
   uint32_t size;
   uint32_t version;
};
*/

typedef struct _progdata progdata;
struct _progdata {
   int verbose;
   int listinfo;
   int listfunc;
   char *pluginfile;
   void *pluginhandle;
   int bapiversion;
   int bplugtype;
   pluginfo *pinfo;
   plugfuncs *pfuncs;
};

/* memory allocation/deallocation */
#define MALLOC(size) \
   (char *) bmalloc ( size );

#define ASSERT_MEMORY(m) \
   if ( m == NULL ){ \
      printf ( "Error: memory allocation error!\n" ); \
      exit (10); \
   }

#define FREE(ptr) \
   if ( ptr != NULL ){ \
      bfree ( ptr ); \
      ptr = NULL; \
   }

int registerBaculaEvents(void *ctx, ...)
{
   return 0;
};

int getBaculaValue(void *ctx, int var, void *value)
{
   return 0;
};

int setBaculaValue(void *ctx, int var, void *value)
{
   return 0;
};

int DebugMessage(void *ctx, const char *file, int line, int level, const char *fmt, ...)
{
#ifdef DEBUGMSG
   printf("DG: %s:%d %s\n", file, line, fmt);
#endif
   return 0;
};

int JobMessage(void *ctx, const char *file, int line, int type, int64_t mtime,
          const char *fmt, ...)
{
#ifdef DEBUGMSG
   printf("JM: %s:%d <%d> %s\n", file, line, type, fmt);
#endif
   return 0;
};

void *baculaMalloc(void *ctx, const char *file, int line, size_t size)
{
   return MALLOC(size);
};

void baculaFree(void *ctx, const char *file, int line, void *mem)
{
   FREE(mem);
};

/*
 * displays a short help
 */
void print_help(int argc, char *argv[])
{

   printf("\n"
     "Usage: bpluginfo [options] <plugin_file.so>\n"
     "       -v          verbose\n"
     "       -i          list plugin header information only (default)\n"
     "       -f          list plugin functions information only\n"
     "       -a <api>    bacula api version (default %d)\n"
     "       -h          help screen\n" "\n", DEFAULT_API_VERSION);
}

/* allocates and resets a main program data variable */
progdata *allocpdata(void)
{

   progdata *pdata;

   pdata = (progdata *) bmalloc(sizeof(progdata));
   ASSERT_MEMORY(pdata);
   memset(pdata, 0, sizeof(progdata));

   return pdata;
}

/* releases all allocated program data resources */
void freepdata(progdata * pdata)
{

   if (pdata->pluginfile) {
      FREE(pdata->pluginfile);
   }
   FREE(pdata);
}

/*
 * parse execution arguments and fills required pdata structure fields
 *
 * input:
 *    pdata - pointer to program data structure
 *    argc, argv - execution envinroment variables
 * output:
 *    pdata - required structure fields
 *
 * supported options:
 * -v    verbose flag
 * -i    list plugin header info only (default)
 * -f    list implemented functions only
 * -a    bacula api version (default 1)
 * -h    help screen
 */
void parse_args(progdata * pdata, int argc, char *argv[])
{

   int i;
   char *dirtmp;
   char *progdir;
   int api;
   int s;

   if (argc < 2) {
      /* TODO - add a real help screen */
      printf("\nNot enough parameters!\n");
      print_help(argc, argv);
      exit(1);
   }

   if (argc > 5) {
      /* TODO - add a real help screen */
      printf("\nToo many parameters!\n");
      print_help(argc, argv);
      exit(1);
   }

   for (i = 1; i < argc; i++) {
      if (strcmp(argv[i], "-h") == 0) {
         /* help screen */
         print_help(argc, argv);
         exit(0);
      }
      if (strcmp(argv[i], "-v") == 0) {
          /* verbose option */
          pdata->verbose = 1;
          continue;
      }
      if (strcmp(argv[i], "-f") == 0) {
         /* functions list */
         pdata->listfunc = 1;
         continue;
      }
      if (strcmp(argv[i], "-i") == 0) {
         /* header list */
         pdata->listinfo = 1;
         continue;
      }
      if (strcmp(argv[i], "-a") == 0) {
         /* bacula api version */
         if (i < argc - 1) {
            s = sscanf(argv[i + 1], "%d", &api);
            if (s == 1) {
               pdata->bapiversion = api;
               i++;
               continue;
            }
         }
         printf("\nAPI version number required!\n");
         print_help(argc, argv);
         exit(1);
      }
      if (!pdata->pluginfile) {
          if (argv[i][0] != '/') {
             dirtmp = MALLOC(PATH_MAX);
             ASSERT_MEMORY(dirtmp);
             progdir = MALLOC(PATH_MAX);
             ASSERT_MEMORY(progdir);
             dirtmp = getcwd(dirtmp, PATH_MAX);

             strcat(dirtmp, "/");
             strcat(dirtmp, argv[i]);

             if (realpath(dirtmp, progdir) == NULL) {
                /* error in resolving path */
                FREE(progdir);
                progdir = bstrdup(argv[i]);
             }
             pdata->pluginfile = bstrdup(progdir);
             FREE(dirtmp);
             FREE(progdir);
          } else {
             pdata->pluginfile = bstrdup(argv[i]);
          }
    continue;
      }
   }
}

/*
 * checks a plugin type based on a plugin magic string
 *
 * input:
 *    pdata - program data with plugin info structure
 * output:
 *    int - enum plugintype
 */
int getplugintype(progdata * pdata)
{

   ASSERT_NVAL_RET_V(pdata, ERRORPLUGIN);

   pluginfo *pinfo = pdata->pinfo;

   ASSERT_NVAL_RET_V(pinfo, ERRORPLUGIN);

   if (pinfo->pdirinfo.plugin_magic &&
       strcmp(pinfo->pdirinfo.plugin_magic, DIR_PLUGIN_MAGIC) == 0) {
      return DIRPLUGIN;
   } else
      if (pinfo->pfdinfo.plugin_magic &&
     strcmp(pinfo->pfdinfo.plugin_magic, FD_PLUGIN_MAGIC) == 0) {
      return FDPLUGIN;
   } else
      if (pinfo->psdinfo.plugin_magic &&
     strcmp(pinfo->psdinfo.plugin_magic, SD_PLUGIN_MAGIC) == 0) {
      return SDPLUGIN;
   } else {
      return ERRORPLUGIN;
   }
}

/*
 * prints any available information about a plugin
 *
 * input:
 *    pdata - program data with plugin info structure
 * output:
 *    printed info
 */
void dump_pluginfo(progdata * pdata)
{

   ASSERT_NVAL_RET(pdata);

   pluginfo *pinfo = pdata->pinfo;

   ASSERT_NVAL_RET(pinfo);

   plugfuncs *pfuncs = pdata->pfuncs;

   ASSERT_NVAL_RET(pfuncs);

   switch (pdata->bplugtype) {
   case DIRPLUGIN:
      printf("\nPlugin type:\t\tBacula Director plugin\n");
      if (pdata->verbose) {
         printf("Plugin magic:\t\t%s\n", NPRT(pinfo->pdirinfo.plugin_magic));
      }
      printf("Plugin version:\t\t%s\n", pinfo->pdirinfo.plugin_version);
      printf("Plugin release date:\t%s\n", NPRT(pinfo->pdirinfo.plugin_date));
      printf("Plugin author:\t\t%s\n", NPRT(pinfo->pdirinfo.plugin_author));
      printf("Plugin licence:\t\t%s\n", NPRT(pinfo->pdirinfo.plugin_license));
      printf("Plugin description:\t%s\n", NPRT(pinfo->pdirinfo.plugin_description));
      printf("Plugin API version:\t%d\n", pinfo->pdirinfo.version);
      break;
   case FDPLUGIN:
      printf("\nPlugin type:\t\tFile Daemon plugin\n");
      if (pdata->verbose) {
         printf("Plugin magic:\t\t%s\n", NPRT(pinfo->pfdinfo.plugin_magic));
      }
      printf("Plugin version:\t\t%s\n", pinfo->pfdinfo.plugin_version);
      printf("Plugin release date:\t%s\n", NPRT(pinfo->pfdinfo.plugin_date));
      printf("Plugin author:\t\t%s\n", NPRT(pinfo->pfdinfo.plugin_author));
      printf("Plugin licence:\t\t%s\n", NPRT(pinfo->pfdinfo.plugin_license));
      printf("Plugin description:\t%s\n", NPRT(pinfo->pfdinfo.plugin_description));
      printf("Plugin API version:\t%d\n", pinfo->pfdinfo.version);
      break;
   case SDPLUGIN:
      printf("\nPlugin type:\t\tBacula Storage plugin\n");
      if (pdata->verbose) {
         printf("Plugin magic:\t\t%s\n", NPRT(pinfo->psdinfo.plugin_magic));
      }
      printf("Plugin version:\t\t%s\n", pinfo->psdinfo.plugin_version);
      printf("Plugin release date:\t%s\n", NPRT(pinfo->psdinfo.plugin_date));
      printf("Plugin author:\t\t%s\n", NPRT(pinfo->psdinfo.plugin_author));
      printf("Plugin licence:\t\t%s\n", NPRT(pinfo->psdinfo.plugin_license));
      printf("Plugin description:\t%s\n", NPRT(pinfo->psdinfo.plugin_description));
      printf("Plugin API version:\t%d\n", pinfo->psdinfo.version);
      break;
   default:
      printf("\nUnknown plugin type or other Error\n\n");
   }
}

/*
 * prints any available information about plugin' functions
 *
 * input:
 *    pdata - program data with plugin info structure
 * output:
 *    printed info
 */
void dump_plugfuncs(progdata * pdata)
{

   ASSERT_NVAL_RET(pdata);

   plugfuncs *pfuncs = pdata->pfuncs;

   ASSERT_NVAL_RET(pfuncs);

   printf("\nPlugin functions:\n");

   switch (pdata->bplugtype) {
   case DIRPLUGIN:
      if (pdata->verbose) {
          if (pfuncs->pdirfuncs.newPlugin) {
             printf(" newPlugin()\n");
          }
          if (pfuncs->pdirfuncs.freePlugin) {
             printf(" freePlugin()\n");
          }
      }
      if (pfuncs->pdirfuncs.getPluginValue) {
         printf(" getPluginValue()\n");
      }
      if (pfuncs->pdirfuncs.setPluginValue) {
         printf(" setPluginValue()\n");
      }
      if (pfuncs->pdirfuncs.handlePluginEvent) {
         printf(" handlePluginEvent()\n");
      }
      break;
   case FDPLUGIN:
      if (pdata->verbose) {
          if (pfuncs->pfdfuncs.newPlugin) {
             printf(" newPlugin()\n");
          }
          if (pfuncs->pfdfuncs.freePlugin) {
             printf(" freePlugin()\n");
          }
      }
      if (pfuncs->pfdfuncs.getPluginValue) {
         printf(" getPluginValue()\n");
      }
      if (pfuncs->pfdfuncs.setPluginValue) {
         printf(" setPluginValue()\n");
      }
      if (pfuncs->pfdfuncs.handlePluginEvent) {
         printf(" handlePluginEvent()\n");
      }
      if (pfuncs->pfdfuncs.startBackupFile) {
         printf(" startBackupFile()\n");
      }
      if (pfuncs->pfdfuncs.endBackupFile) {
         printf(" endBackupFile()\n");
      }
      if (pfuncs->pfdfuncs.startRestoreFile) {
         printf(" startRestoreFile()\n");
      }
      if (pfuncs->pfdfuncs.endRestoreFile) {
         printf(" endRestoreFile()\n");
      }
      if (pfuncs->pfdfuncs.pluginIO) {
         printf(" pluginIO()\n");
      }
      if (pfuncs->pfdfuncs.createFile) {
         printf(" createFile()\n");
      }
      if (pfuncs->pfdfuncs.setFileAttributes) {
         printf(" setFileAttributes()\n");
      }
      if (pfuncs->pfdfuncs.checkFile) {
         printf(" checkFile()\n");
      }
      break;
   case SDPLUGIN:
      if (pdata->verbose) {
          if (pfuncs->psdfuncs.newPlugin) {
             printf(" newPlugin()\n");
          }
          if (pfuncs->psdfuncs.freePlugin) {
             printf(" freePlugin()\n");
          }
      }
      if (pfuncs->psdfuncs.getPluginValue) {
         printf(" getPluginValue()\n");
      }
      if (pfuncs->psdfuncs.setPluginValue) {
         printf(" setPluginValue()\n");
      }
      if (pfuncs->psdfuncs.handlePluginEvent) {
         printf(" handlePluginEvent()\n");
      }
      break;
   default:
      printf("\nUnknown plugin type or other Error\n\n");
   }
}

/*
 * input parameters:
 *    argv[0] [options] <plugin_filename.so>
 *
 * exit codes:
 *    0 - success
 *    1 - cannot load a plugin
 *    2 - cannot find a loadPlugin function
 *    3 - cannot find an unloadPlugin function
 *    10 - not enough memory
 */
int main(int argc, char *argv[])
{

   progdata *pdata;
   loadPlugin loadplugfunc;
   unloadPlugin unloadplugfunc;
   baculafuncs bfuncs = {
      sizeof(bfuncs),
      1,
      registerBaculaEvents,
      getBaculaValue,
      setBaculaValue,
      JobMessage,
      DebugMessage,
      baculaMalloc,
      baculaFree,
   };
   baculainfos binfos;

   pdata = allocpdata();
   parse_args(pdata, argc, argv);

   binfos.bfdinfo.size = sizeof(binfos);
   binfos.bfdinfo.version = DEFAULT_API_VERSION;

   pdata->pluginhandle = dlopen(pdata->pluginfile, RTLD_LAZY);
   if (pdata->pluginhandle == NULL) {
      printf("\nCannot load a plugin: %s\n\n", dlerror());
      freepdata(pdata);
      exit(1);
   }

   loadplugfunc = (loadPlugin) dlsym(pdata->pluginhandle, "loadPlugin");
   if (loadplugfunc == NULL) {
      printf("\nCannot find loadPlugin function: %s\n", dlerror());
      printf("\nWhether the file is a really Bacula plugin?\n\n");
      freepdata(pdata);
      exit(2);
   }

   unloadplugfunc = (unloadPlugin) dlsym(pdata->pluginhandle, "unloadPlugin");
   if (unloadplugfunc == NULL) {
      printf("\nCannot find unloadPlugin function: %s\n", dlerror());
      printf("\nWhether the file is a really Bacula plugin?\n\n");
      freepdata(pdata);
      exit(3);
   }

   if (pdata->bapiversion > 0) {
      binfos.bdirinfo.version = pdata->bapiversion;
   }

   loadplugfunc(&binfos, &bfuncs, (void **)&pdata->pinfo, (void **)&pdata->pfuncs);

   pdata->bplugtype = getplugintype(pdata);

   if (!pdata->listfunc) {
      dump_pluginfo(pdata);
   }
   if ((!pdata->listinfo && pdata->listfunc) || pdata->verbose) {
      dump_plugfuncs(pdata);
   }
   printf("\n");

   unloadplugfunc();

   dlclose(pdata->pluginhandle);

   freepdata(pdata);

   return 0;
}
